File: src/main/java/org/veupathdb/lib/blast/BlastConfig.java
package org.veupathdb.lib.blast;

import org.veupathdb.lib.blast.consts.Flag;
import org.veupathdb.lib.blast.consts.Key;
import org.veupathdb.lib.blast.field.*;
import org.veupathdb.lib.blast.util.JSONDecodable;
import org.veupathdb.lib.blast.util.JSONEncodable;
import org.veupathdb.lib.blast.util.JSONObjectDecoder;
import org.veupathdb.lib.blast.util.JSONObjectEncoder;

public interface BlastConfig extends JSONEncodable, JSONDecodable
{
  BlastTool getTool();

  Boolean getShortHelp();

  void setShortHelp(Boolean val);

  Boolean getLongHelp();

  void setLongHelp(Boolean val);

  Boolean getVersion();

  void setVersion(Boolean val);

  OutFile getOutFile();

  void setOutFile(String file);

  OutFormat getOutFormat();

  void setOutFormat(OutFormat fmt);

  Boolean getShowGIs();

  void setShowGIs(Boolean val);

  NumDescriptions getNumDescriptions();

  void setNumDescriptions(Long val);

  NumAlignments getNumAlignments();

  void setNumAlignments(Long val);

  LineLength getLineLength();

  void setLineLength(Integer val);

  Boolean getHTML();

  void setHTML(Boolean val);

  HitSorting getSortHits();

  void setSortHits(HitSorting val);

  HSPSorting getSortHSPs();

  void setSortHSPs(HSPSorting val);

  MaxTargetSeqs getMaxTargetSequences();

  void setMaxTargetSequences(Long val);

  Boolean getParseDefLines();

  void setParseDefLines(Boolean val);

  default JSONObjectEncoder toJSON(boolean includeTool) {
    final var out = new JSONObjectEncoder();

    if (includeTool)
      out.encode(Key.Tool, getTool());

    out.encode(Flag.ShortHelp, getShortHelp());
    out.encode(Flag.LongHelp, getLongHelp());
    out.encode(Flag.Version, getVersion());
    out.encode(Flag.OutFile, getOutFile());
    out.encode(Flag.OutFormat, getOutFormat());
    out.encode(Flag.ShowGIs, getShowGIs());
    out.encode(Flag.NumDescriptions, getNumDescriptions());
    out.encode(Flag.NumAlignments, getNumAlignments());
    out.encode(Flag.LineLength, getLineLength());
    out.encode(Flag.HTML, getHTML());
    out.encode(Flag.SortHits, getSortHits());
    out.encode(Flag.SortHSPs, getSortHSPs());
    out.encode(Flag.MaxTargetSequences, getMaxTargetSequences());
    out.encode(Flag.ParseDefLines, getParseDefLines());

    return out;
  }

  @Override
  default JSONObjectEncoder toJSON() {
    return toJSON(false);
  }

  @Override
  default void decodeJSON(JSONObjectDecoder node) {
    node.decode(Flag.ShortHelp, this::setShortHelp);
    node.decode(Flag.LongHelp, this::setLongHelp);
    node.decode(Flag.Version, this::setVersion);
    node.decode(Flag.OutFile, this::setOutFile);
    node.decode(Flag.OutFormat, this::setOutFormat, OutFormat::fromJSON);
    node.decode(Flag.ShowGIs, this::setShowGIs);
    node.decode(Flag.NumDescriptions, this::setNumDescriptions);
    node.decode(Flag.NumAlignments, this::setNumAlignments);
    node.decode(Flag.LineLength, this::setLineLength);
    node.decode(Flag.HTML, this::setHTML);
    node.decode(Flag.SortHits, this::setSortHits, HitSorting::fromJSON);
    node.decode(Flag.SortHSPs, this::setSortHSPs, HSPSorting::fromJSON);
    node.decode(Flag.MaxTargetSequences, this::setMaxTargetSequences);
    node.decode(Flag.ParseDefLines, this::setParseDefLines);
  }
}

File: src/main/java/org/veupathdb/lib/blast/BlastConfig.kt
package org.veupathdb.lib.blast

import com.fasterxml.jackson.annotation.JsonValue
import org.veupathdb.lib.blast.consts.Flag
import org.veupathdb.lib.blast.consts.Key
import org.veupathdb.lib.blast.field.*
import org.veupathdb.lib.blast.util.JSONDecodable
import org.veupathdb.lib.blast.util.JSONEncodable
import org.veupathdb.lib.blast.util.JSONObjectDecoder
import org.veupathdb.lib.blast.util.JSONObjectEncoder

sealed interface BlastConfig : JSONEncodable, JSONDecodable
{
  val tool: BlastTool

  var shortHelp: Boolean?

  var longHelp: Boolean?

  var version: Boolean?

  var outFile: OutFile?

  var outFormat: OutFormat?

  var showGIs: Boolean?

  var numDescriptions: NumDescriptions?

  var numAlignments: NumAlignments?

  var lineLength: LineLength?

  var html: Boolean?

  var sortHits: HitSorting?

  var sortHSPs: HSPSorting?

  var maxTargetSequences: MaxTargetSeqs?

  var parseDefLines: Boolean?

  @JsonValue
  override fun toJSON(): JSONObjectEncoder {
    return toJSON(false)
  }

  fun toJSON(includeTool: Boolean): JSONObjectEncoder {
    val out = JSONObjectEncoder()

    if (includeTool)
      out.encode(Key.Tool, tool)

    out.encode(Flag.ShortHelp, shortHelp)
    out.encode(Flag.LongHelp, longHelp)
    out.encode(Flag.Version, version)
    out.encode(Flag.OutFile, outFile)
    out.encode(Flag.OutFormat, outFormat)
    out.encode(Flag.ShowGIs, showGIs)
    out.encode(Flag.NumDescriptions, numDescriptions)
    out.encode(Flag.NumAlignments, numAlignments)
    out.encode(Flag.LineLength, lineLength)
    out.encode(Flag.HTML, html)
    out.encode(Flag.SortHits, sortHits)
    out.encode(Flag.SortHSPs, sortHSPs)
    out.encode(Flag.MaxTargetSequences, maxTargetSequences)
    out.encode(Flag.ParseDefLines, parseDefLines)

    return out
  }

  override fun decodeJSON(node: JSONObjectDecoder) {
    node.decodeBool(Flag.ShortHelp)          { shortHelp          = it                      }
    node.decodeBool(Flag.LongHelp)           { longHelp           = it                      }
    node.decodeBool(Flag.Version)            { version            = it                      }
    node.decodeString(Flag.OutFile)          { outFile            = OutFile(it)             }
    node.decodeJSON(Flag.OutFormat)          { outFormat          = OutFormat.fromJSON(it)  }
    node.decodeBool(Flag.ShowGIs)            { showGIs            = it                      }
    node.decodeLong(Flag.NumDescriptions)    { numDescriptions    = NumDescriptions(it)     }
    node.decodeLong(Flag.NumAlignments)      { numAlignments      = NumAlignments(it)       }
    node.decodeInt(Flag.LineLength)          { lineLength         = LineLength(it)          }
    node.decodeBool(Flag.HTML)               { html               = it                      }
    node.decodeJSON(Flag.SortHits)           { sortHits           = HitSorting.fromJSON(it) }
    node.decodeJSON(Flag.SortHSPs)           { sortHSPs           = HSPSorting.fromJSON(it) }
    node.decodeLong(Flag.MaxTargetSequences) { maxTargetSequences = MaxTargetSeqs(it)       }
    node.decodeBool(Flag.ParseDefLines)      { parseDefLines      = it                      }
  }
}
